### PR TITLE
Add ability to send extended-sys-state message

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

**Comment by peterbarker** [comment]
On Wed, 6 Jun 2018, Pierre Kancir wrote:

> Don't we need to add it to a stream ?

Well I don't - it comes out when you request it when it's on top of the 
message-interval branch ;-)

File: ArduCopter/mode_rtl.cpp [hunk 1]
```diff
@@ -335,6 +335,18 @@ void Copter::ModeRTL::land_start()
     auto_yaw.set_mode(AUTO_YAW_HOLD);
 }
 
+bool Copter::ModeRTL::is_landing() const
+{
+    switch(_state) {
+    case RTL_Land:
+    case RTL_FinalDescent:
```
Comment (OXINARF) [inline ArduCopter/mode_rtl.cpp]: This is a good question, is final descent a landing?

File: ArduPlane/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -1739,3 +1739,37 @@ const AP_FWVersion &GCS_MAVLINK_Plane::get_fwver() const
...
+        case QuadPlane::TRANSITION_DONE:
+            if (plane.quadplane.in_vtol_mode()) {
+                return MAV_VTOL_STATE_MC;
+            } else {
+                return MAV_VTOL_STATE_FW;
+            }
+        }
+        return MAV_VTOL_STATE_UNDEFINED;
+    }
+    return MAV_VTOL_STATE_UNDEFINED;
```
Comment (OXINARF) [inline ArduPlane/GCS_Mavlink.cpp]: Shouldn't this one be MAV_VTOL_STATE_FW?

**Comment by peterbarker** [comment]
On Wed, 6 Jun 2018, Francisco Ferreira wrote:

> I kind of agree with @khancyr on adding this to some stream. I know that
> with message interval this can be asked individually, but we are keeping the
> stream API for now so I think it makes sense to add it to a stream.

I'd suggest this goes in after the message-intervals patch.

If we add this in now and add it to a stream then we can't take it out of 
that stream later - or we may break GCSs which have started to use it but 
aren't doing the set-message-interval thing yet.

In fact, going forward, I think we shouldn't add most messages to streams; 
if a GCS is capable of consuming a message it should request it.

> +{
> +    switch(_state) {
> +    case RTL_Land:
> +    case RTL_FinalDescent:
> 
> This is a good question, is final descent a landing?

I think so....

> +    return MAV_VTOL_STATE_UNDEFINED;
> Shouldn't this one be MAV_VTOL_STATE_FW?

Not by my reading of the standard:

https://github.com/ArduPilot/mavlink/blob/master/message_definitions/v1.0/common.xml#L2425

Not saying it's a *good* standard, but it does give you more information, 
I guess....

**Comment by OXINARF** [comment]
> If we add this in now and add it to a stream then we can't take it out of that stream later - or we may break GCSs which have started to use it but aren't doing the set-message-interval thing yet.

@peterbarker I agree that if at any point we add it to a stream then we shouldn't remove it. It just seems to me that it would make sense to have this extra information in one.

> I think so....

@peterbarker Final descent isn't going to land, just descend, which is why I'm not sure that qualifies as landing - that would mean that any descend in any mode would also be a landing.

> Not by my reading of the standard:
>
>https://github.com/ArduPilot/mavlink/blob/master/message_definitions/v1.0/common.xml#L2425
>
>Not saying it's a *good* standard, but it does give you more information, I guess....

@peterbarker OK, makes sense.

**Comment by peterbarker** [comment]
Randy indicated in the devcall that final descent should be considered LAN.

This issue is now dependant on the message-intervals patch (https://github.com/ArduPilot/ardupilot/pull/7234) going in.

**Comment by peterbarker** [comment]
(it was agreed on the call this could be merged as soon as it works on top of message-intervals)

**Comment by khancyr** [comment]
need rebase and squash some change and ready to go in as we have message interval!

**Comment by peterbarker** [comment]
@khancyr So it turns out quadplane wasn't recording state in a way which made it nice to report on.

I've taken a stab at fixing that.

I've also added a test suite for the message.

File: ArduCopter/mode_rtl.cpp (hunk 1 above)
Comment (khancyr) [reply ArduCopter/mode_rtl.cpp]: @peterbarker could you put RTL_FinalDescent in no landing ? after testing this in my company, we found this is better to not beeing flagged as landing

**Comment by peterbarker** [comment]
@khancyr Changed it so finaldescent isn't considered landing.  Just like @OXINARF told me :-)

File: ArduCopter/mode_rtl.cpp (hunk 1 above)
Comment (peterbarker) [reply ArduCopter/mode_rtl.cpp]: Done.

**Comment by peterbarker** [comment]
I'm going to split the Plane support out into a different pull request.  Those changes are not simple mavlink-message changes, and my guess is they're going to sit around for a long time.

I'm also going to separate split out the is-flying changes.

**Comment by rmackay9** [comment]
Merged, thanks!

**Pull request merged**